File: TimeSeries/PublicApis/R/timeseries_client.R
# Load dependencies
require('jsonlite')
require('httr')

# Sys.setenv(http_proxy="http://localhost:8888") # Enables Fiddler capturing of traffic
# Sys.setenv(http_proxy="") # Disables Fiddler proxying

# Create a simple AQUARIUS Time-Series API client.
timeseriesClient <- setRefClass("timeseriesClient",
  fields = list(version = "character", publishUri = "character", acquisitionUri = "character", provisioningUri = "character"),
  methods = list(
#' Connects to an AQTS server
#' 
#' Once authenticated, all subsequent requests to the AQTS server will reuse the authenticated session
#' 
#' @param hostname A server name or IP address
#' @param username The AQTS credentials username
#' @param password The AQTS credentials password
#' @examples 
#' connect("localhost", "admin", "admin") # When running R on your AQTS app server
#' connect("myserver", "me", "mypassword") # Connect over the network
#' connect("https://myserver", "user", "letmein") # Connect to an AQTS server with HTTPS enabled
    connect = function(hostname, username, password) {
      # Auto-configure the proxy by default
      .self$configureProxy()

      # Support schemeless and schemed hosts for convenience
      prefix <- "http://"
      if (startsWith(hostname, "http://") || startsWith(hostname, "https://")) {
        url <- parse_url(hostname)
        hostname <- paste0(url$scheme, "://", url$hostname)
        prefix <- ""
      }

      # Grab the version of the AQTS server
      r <- GET(paste0(prefix, hostname, "/AQUARIUS/apps/v1/version"))
      stop_for_status(r, "detecting AQTS version")

      j <- fromJSON(content(r, "text"))
      version <<- j$ApiVersion

      # Compose the base URI for all API endpoints
      publishUri <<- paste0(prefix, hostname, "/AQUARIUS/Publish/v2")
      acquisitionUri <<- paste0(prefix, hostname, "/AQUARIUS/Acquisition/v2")
      provisioningUri <<- paste0(prefix, hostname, "/AQUARIUS/Provisioning/v1")

      # Try to authenticate using the supplied credentials
      r <- POST(paste0(publishUri, "/session"), body = list(Username = username, EncryptedPassword = password), encode = "json")
      stop_for_status(r, "authenticate with AQTS")
    },

#' Disconnects immediately from an AQTS server
    disconnect = function() {
      r <- DELETE(paste0(publishUri, "/session"))
      stop_for_status(r, "disconnect from AQTS")
    },

#' Auto-configures the proxy to route all requests through Fiddler
#'
#' This method configures the R proxy to route everything through Fiddler if it is running
#' 
#' Sys.setenv(http_proxy="http://localhost:8888") # Enables Fiddler capturing of traffic
#' Sys.setenv(http_proxy="") # Disables Fiddler proxying
    configureProxy = function() {
      if (length(Sys.getenv("R_DISABLE_FIDDLER")[0]) > 0 || Sys.info()['sysname'] != "Windows")
        return()

      # Check if Fiddler is running
      taskCheck <- system('tasklist /FI "IMAGENAME eq Fiddler.exe"', intern = TRUE)

      if (length(taskCheck) < 4)
        return()

      # Fiddler is running, so enable the proxy
      message("Auto-routing all web requests through Fiddler.")
      message('To disable this behaviour, call Sys.setenv(http_proxy="") or set the R_DISABLE_FIDDLER environment variable.')
      Sys.setenv(http_proxy = "http://localhost:8888")
    },

#' Determines if a target version string is strictly less than a source version
#' 
#' This method takes dotted version strings and compares them by numerical components.
#' It safely avoids the errors string comparison, which incorrectly says "3.10.510" > "17.2.123".
#' @param targetVersion Target version string
#' @param sourceVersion Optional source version string. If missing, use the connected server version
#' @return TRUE if the target version is strictly less than the source version
    isVersionLessThan = function(targetVersion, sourceVersion) {
      if (missing(sourceVersion)) {
        sourceVersion <- version
      }

      # Create the vectors of integers using this local sanitizing function
      createIntegerVector <- function(versionText) {

        if (versionText == "0.0.0.0") {
          # Force unreleased developer builds to act as latest-n-greatest
          versionText <- "9999.99"
        }

        # Convert the text into a vector of integers
        v <- as.integer(strsplit(versionText, ".", fixed = TRUE)[[1]])

        if (length(v) > 0 && v[1] >= 14 && v[1] <= 99) {
          # Adjust the leading component to match the 20xx.y release convention
          v[1] = v[1] + 2000
        }

        v
      }

      # Convert to vectors of integers
      target <- createIntegerVector(targetVersion)
      source <- createIntegerVector(sourceVersion)

      # Take the differnce of the common parts
      minlength <- min(length(target), length(source))

      diff <- head(target, minlength) - head(source, minlength)

      if (all(diff == 0)) {
        # All the common parts are identical
        length(source) < length(target)
      } else {
        # Assume not less than
        lessThan <- FALSE

        for (d in diff) {
          if (d < 0) {
            break
          } else if (d > 0) {
            lessThan <- TRUE
            break
          }
        }

        lessThan
      }
    },

#' Gets the unique ID of a time-series from its identifier string
#' 
#' If the input string is already a unique ID, the input value is simply returned unmodified.
#' 
#' @param timeSeriesIdentifier A time-series identifier in <Parameter>.<Label>@<LocationIdentifier> syntax
#' @return The unique ID of the time-series
#' @examples
#' getTimeSeriesUniqueId("Stage.Working@MyLocation") # cdf184928c8249abb872f852f0fa7d01
    getTimeSeriesUniqueId = function(timeSeriesIdentifier) {
      if (!grepl("@", timeSeriesIdentifier)) {
        # It's not in Param.Label@Location format, so just leave it as-is
        timeSeriesIdentifier
      } else {
        # Parse out the location identifier
        location <- .self$getLocationIdentifier(timeSeriesIdentifier)

        # Ask for all the time-series at that location
        r <- GET(paste0(publishUri, "/GetTimeSeriesDescriptionList"), query = list(LocationIdentifier = location))
        stop_for_status(r, paste("retrieve time-series at location", location))

        # Find the unique ID by matching the full identifier
        j <- fromJSON(content(r, "text"))
        j$TimeSeries$UniqueId[which(j$TimeSeries$Identifier == timeSeriesIdentifier)]
      }
    },

#' Gets the location identifier from a time series identifier string
#'
#' @param timeSeriesIdentifier A time-series identifier in <Parameter>.<Label>@<LocationIdentifier> syntax
#' @return The identifier of the location
#' @examples
#' getLocationIdentifier("Stage.Working@MyLocation") # MyLocation
    getLocationIdentifier = function(timeSeriesIdentifier) {
      if (!grepl("@", timeSeriesIdentifier)) {
        stop(timeSeriesIdentifier, " is not a <Parameter>.<Label>@<Location> time-series identifier")
      }

      strsplit(timeSeriesIdentifier, "@")[[1]][[2]]
    },

#' Parse an ISO 8601 timestamp into a POSIXct value
#' 
#' @param isoText An ISO 8601 timestamp string
#' @return The equivalent POSIXct datetime
#' @examples 
#' parseIso8601("2015-04-01T00:00:00Z") # April Fool's day, 2015 UTC
#' parseIso8601("2015-04-01T00:00:00-08:00") # April Fool's day, 2015, Pacific Standard Time
#'
#' times = sapply(json$Points$Timestamp, timeseries$parse8601) # Convert all JSON timestamp strings into POSIXct format
    parseIso8601 = function(isoText) {
      # Wow. Parsing true ISO 8061 timestamps (which AQTS outputs) in R is hard.
      #
      # Parsing them **efficiently** and correctly when dealing with thousands of points is even harder!
      #
      # There are many packages claiming to be compliant with ISO 8601.
      # But most are only compliant-ish(TM), failing to deal with timezones and UTC offsets
      #
      # R's strptime() method has a %z conversion specifier for timezone offsets.
      #
      # But %z only suports RFC 822 offset in: <sign><4-digits>
      #
      #  +HHMM (+1400 max)
      #  -HHMM (-1400 max)
      #
      # AQTS uses ISO 8601 offsets, either "Z" or <sign><2-digit-hour>:<2-digit-minute>
      #
      #  Z      (UTC assumed)
      #  +HH:MM (+14:00 max)
      #  -HH:MM (-12:00 min)
      #
      # This function can process roughly 10K timestamps/sec.
      # By comparison, the popular-and-otherwise-correct lubridate library is 60x slower at ~ 150 timestamps per second
      len <- nchar(isoText)

      if (substr(isoText, len - 2, len - 2) == ":") {
        # The most common scenario from AQTS output: A truly correct ISO 8601 timestamp with a numeric UTC offset
        # Strip out the colon separating the UTC offset, since that is what %z requires
        isoText <- paste0(substr(isoText, 1, len - 3), substr(isoText, len - 1, len))
      } else if (substr(isoText, len, len) == "Z") {
        # Second most likely scenario from AQTS output: The "Z" representing a UTC time
        # Convert the unsupported UTC shorthand into an offset with no effect
        isoText <- paste0(substr(isoText, 1, len - 1), "+0000")
      }

      as.POSIXct(strptime(isoText, "%Y-%m-%dT%H:%M:%OS%z", "UTC"))
    },

#' Formats a datetime in ISO 8601 format
#' 
#' @param datetime A datetime object
#' @return The time in YYYY-MM-DDTHH:mm:SS.fffffZ format
    formatIso8601 = function(datetime) {
      isoText <- strftime(datetime, "%Y-%m-%dT%H:%M:%OS%z", "UTC")
      
      len <- nchar(isoText)
      
      if (substr(isoText, len, len) != "Z") {
        # Inject the missing colon in the zone offset, so "+HHMM" becomes "+HH:MM"
        isoText = paste0(substr(isoText, 1, len - 2), ":", substr(isoText, len - 1, len))
      }
      
      isoText
    },

#' Gets the UTC offset string from a numeric UTC offset hours value
#'
#' @param utcOffset A UTC offset in hours
#' @return An ISO8601 UTC offset string in +/-HH:MM format
#' @examples
#' getUtcOffsetText(-8) # "-08:00"
#' getUtcOffsetText(2.5) # "+02:30"
    getUtcOffsetText = function(utcOffset) {
      isNegative <- FALSE
      totalMinutes <- as.integer(utcOffset * 60)

      if (totalMinutes < 0) {
        isNegative = TRUE
        totalMinutes = -totalMinutes
      }

      hours = as.integer(totalMinutes / 60)
      minutes = as.integer(totalMinutes %% 60)

      sprintf("%s%02d:%02d", if (isNegative) "-" else "+", hours, minutes)
    },

#' Gets the location data for a location
    getLocationData = function(locationIdentifier) {
      r <- GET(paste0(publishUri, "/GetLocationData"), query = list(LocationIdentifier = locationIdentifier))
      stop_for_status(r, paste("get location data for", locationIdentifier))

      j <- fromJSON(content(r, "text"))
    },

#' Gets time-series points for multiple time-series
#' 
#' Retrieves points for up to 10 time-series.
#' Point values from secondary time-series will be time-aligned via interpolation
#' rules to the timestamps from the first time-series.
#' 
#' @param timeSeriesIds A list of time-series identifiers or unique IDs
#' @param queryFrom Optional time from which to retrieve data.If missing, fetches data from the start-of-record
#' @param queryTo Optional time to which data willl be retrieved. If missing, fetches data to the end-of-record
#' @param outputUnitIds Optional unit IDs for output. If missing or empty, the default unit of the time-series will be used
#' @returns The JSON object from the /GetTimeSeriesData response
#' @examples
#' ## Get the discharge and stage timeseries for 2012
#' json = timeseries$getTimeSeriesData(c("Discharge.Working@Location","Stage.Working@Location"),
#'                                     queryFrom = "2012-01-01T00:00:00Z",
#'                                     queryTo   = "2013-01-01T00:00:00Z")
#'
#' ## Plot stage vs dicharge
#' plot(json$Points$NumericValue1, json$Points$NumericValue2)
#'
#' ## Plot stage vs dicharge, with log scale, and some labeled axis
#' plot(json$Points$NumericValue1, json$Points$NumericValue2, log = "xy",
#'      xlab = json$TimeSeries$Identifier[1],
#'      ylab = json$TimeSeries$Identifier[2])
    getTimeSeriesData = function(timeSeriesIds, queryFrom, queryTo, outputUnitIds) {
      if (.self$isVersionLessThan("17.2")) {
        # Throw on the brakes if the server is too old
        stop("Time aligned data is not available before AQTS 2017.2. Connected server version=", version)
      }

      if (is.character(timeSeriesIds)) {
        # Coerce a single timeseries ID string into a vector
        timeSeriesIds = c(timeSeriesIds)
      }

      uniqueIds <- lapply(timeSeriesIds, .self$getTimeSeriesUniqueId)

      q <- list(TimeSeriesUniqueIds = paste(uniqueIds, collapse = ","))

      if (!missing(queryFrom)) {
        if (is.double(queryFrom)) {
          # Coerce to an ISO 8601 string
          queryFrom = .self$formatIso8601(queryFrom)
        }

        q <- c(q, QueryFrom = queryFrom)
      }

      if (!missing(queryTo)) {
        if (is.double(queryTo)) {
          # Coerce to an ISO 8601 string
          queryTo = .self$formatIso8601(queryTo)
        }

        q <- c(q, QueryTo = queryTo)
      }

      if (!missing(outputUnitIds)) {
        if (is.character(outputUnitIds)) {
          # Coerce a single unit ID into a vector
          outputUnitIds = c(outputUnitIds)
        }

        q <- c(q, TimeSeriesOutputUnitIds = paste(outputUnitIds, collapse = ","))
      }

      r <- GET(paste0(publishUri, "/GetTimeSeriesData"), query = q)
      stop_for_status(r, paste("get time-aligned data for", length(uniqueIds), "time-series"))

      j <- fromJSON(content(r, "text"))
    },

#' Uploads a file to a location as an external report
#'
#' Any existing report on AQTS with the same title will be 
#'
#' @param locationDataOrIdentifier Either a location identifier string, or a LocationData object from a previous getLocationData request
#' @param path The path to the file to be uploaded
#' @param title The title of the report to display in AQTS
#' @param deleteDuplicateReports If TRUE or missing, any existing reports with the same title will be deleted before the new report is uploaded
#' @returns The JSON response from a successful upload
  uploadExternalReport = function(locationDataOrIdentifier, path, title, deleteDuplicateReports) {
    if (.self$isVersionLessThan("17.3")) {
      # Throw on the brakes if the server is too old
      stop("Uploading external reports is not available before AQTS 2017.3. Connected server version=", version)
    }
  
    if (missing(title)) {
      stop("A report title is required.")
    }
  
    if (!file.exists(path)) {
      stop("Can't upload non-existent file '", path, "'")
    }
  
    if (missing(deleteDuplicateReports)) {
      # Delete duplicate reports by default
      deleteDuplicateReports = TRUE
    }
  
    locationData <- locationDataOrIdentifier
  
    if (!is.list(locationData) || is.null(locationData$UniqueId)) {
      # We don't know the location unique ID, so look it up from the identifier string
      locationData <- .self$getLocationData(locationDataOrIdentifier)
    }
  
    if (deleteDuplicateReports) {
  
      reports = .self$getReportList()$Reports
  
      for (row in 1:nrow(reports)) {
  
        if (reports[row, "IsTransient"]) {
          # Ignore transient reports
          next
        }
  
        if (reports[row, "Title"] != title) {
          # Ignore permanent reports whose title does not match
          next
        }
  
        .self$deleteReport(reports[row, "ReportUniqueId"])
      }
    }
  
    # Upload the file to the location
    r <- POST(paste0(acquisitionUri, "/locations/", locationData$UniqueId, "/attachments/reports"),
       body = list(uploadedFile = upload_file(path), Title = title))
    stop_for_status(r, paste("upload external report to location", locationData$Identifier))
  
    j <- fromJSON(content(r, "text"))
  
  },

#' Gets a list of reports
  getReportList = function() {
  
    r <- GET(paste0(publishUri, "/GetReportList"))
    stop_for_status(r, "get report list")
  
    j <- fromJSON(content(r, "text"))
  },

#' Deletes a report
  deleteReport = function(reportUniqueId) {
    r <- DELETE(paste0(acquisitionUri, "/attachments/reports/", reportUniqueId))
    stop_for_status(r, paste("delete report", reportUniqueId))
  },

#' Performs a batch of identical operations
#' 
#' This method is useful for requesting large amounts of similar data from AQTS,
#' taking advantage of ServiceStack's support for auto-batched requests.
#' 
#' http://docs.servicestack.net/auto-batched-requests
#' 
#' When you find that a public API only supports a 1-at-a-time approach, and your
#' code needs to request thousands of items, the sendBatchRequests() method is the one to use.
#' 
#' @param endpoint The base REST endpoint
#' @param operationName The name of operation, from the AQTS Metadata page, to perform multiple times. NOT the route, but the operation name.
#' @param requests A collection of individual request objects
#' @param batchSize Optional batch size (defaults to 100 requests per batch)
#' @param verb Optional HTTP verb of the operation (defaults to "GET")
#' @returns A single dataframe containing all the batched responses
#' @examples 
#' # Request info for 3 locations.
#' # Single-request URL is GET /Publish/v2/GetLocationData?LocationIdentifer=loc1
#' # Operation name is "LocationDataServiceRequest"
#' requests = c(list(LocationIdentifier="Loc1"), list(LocationIdentifier="Loc3"), list(LocationIdentifier="Loc3"))
#' responses = timeseries$sendBatchRequests(timeseries$publishUri,"LocationDataServiceRequest", requests)
  sendBatchRequests = function(endpoint, operationName, requests, batchSize, verb) {
    
    if (missing(batchSize)) {
      # Use a reasonable batch size
      batchSize <- 100
    }
    
    if (missing(verb)) {
      verb <- "GET"
    }
    
    # Compose the special batch-operation URL supported by ServiceStack
    url = paste0(endpoint, "/json/reply/", operationName, "[]")
    
    # Split the requests into batche-sized chunks
    batches <- split(requests, ceiling(seq_along(requests) / batchSize))
    
    # Create a local function to request each batch of requests, using the verb as an override to the POST
    batchPost <- function(batches) {
      r <- POST(url, body = batches, encode = "json", add_headers("X-Http-Method-Override" = verb))
      stop_for_status(r, "sending batch requests")
      
      j <- fromJSON(content(r, "text"))
    }
    
    # Call the operation in batches
    batchResponses <- lapply(batches, batchPost)
    
    # Flatten the list of response dataframes into a single data frame
    j <- do.call("rbind", batchResponses)
  }
  )
)

# Create a client in the global namespace
timeseries = timeseriesClient()

File: TimeSeries/PublicApis/Python/timeseries_client.py
# Sample python code to Public API course
# Requires python 2.7+
# Install required dependencies via: $ pip install requests pytz pyrfc3339

import subprocess
import os
import requests
from requests.exceptions import HTTPError
import pyrfc3339
from datetime import datetime
import re


def create_endpoint(hostname, root_path):
    prefix = "http://"
    if hostname.startswith("http://") or hostname.startswith("https://"):
        prefix = ""
    return "{0}{1}{2}".format(prefix, hostname, root_path)


def response_or_raise(response):
    if response.status_code >= 400:
        json = response.json
        if isinstance(json, dict) and json.has_key('ResponseStatus'):
            http_error_msg = u'%s WebService Error: %s(%s) for url: %s' % (
                response.status_code, response.reason, json['ResponseStatus']['Message'], response.url)
            raise HTTPError(http_error_msg, response=response)

    response.raise_for_status()
    return response


class ModelNotFoundException(Exception):
    """Exception raised for errors in the input.

        Attributes:
            identifier -- The identifier that was not found
            message -- explanation of the error
        """

    def __init__(self, identifier, message):
        self.identifier = identifier
        self.message = message


class LocationNotFoundException(ModelNotFoundException):
    """"Raised when the location identifier is not known"""

    def __init__(self, identifier):
        super(ModelNotFoundException, self).__init__(identifier, "Location '{0}' not found.".format(identifier))


class TimeSeriesNotFoundException(ModelNotFoundException):
    """Raised when the time-series identifier cannot be found."""

    def __init__(self, identifier):
        super(ModelNotFoundException, self).__init__(identifier, "Time-series '{0}' not found.".format(identifier))


class TimeseriesSession(requests.sessions.Session):
    """
    A requests.Session object that:
    - Sends all requests to a base endpoint
    - Always raises an exception if any HTTP errors are detected.

    >>> session.get('/invalidroute') # Raises HTTPError (404)
    """

    def __init__(self, hostname, root_path, verify=True):
        super(TimeseriesSession, self).__init__()
        self.verify = verify
        self.base_url = create_endpoint(hostname, root_path)

    def get(self, url, **kwargs):
        r = super(TimeseriesSession, self).get(self.base_url + url, verify=self.verify, **kwargs)
        return response_or_raise(r)

    def post(self, url, data=None, json=None, **kwargs):
        r = super(TimeseriesSession, self).post(self.base_url + url, data, json, verify=self.verify, **kwargs)
        return response_or_raise(r)

    def put(self, url, data=None, **kwargs):
        r = super(TimeseriesSession, self).put(self.base_url + url, data, verify=self.verify, **kwargs)
        return response_or_raise(r)

    def delete(self, url, **kwargs):
        r = super(TimeseriesSession, self).delete(self.base_url + url, verify=self.verify, **kwargs)
        return response_or_raise(r)

    def set_session_token(self, token):
        self.headers.update({"X-Authentication-Token": token})

    def send_batch_requests(self, operation_name, requests, batch_size=100, verb="GET"):
        """
        Performs a batch of identical operations

        This method is useful for requesting large amounts of similar data from AQTS,
        taking advantage of ServiceStack's support for auto-batched requests.

        http://docs.servicestack.net/auto-batched-requests

        When you find that a public API only supports a 1-at-a-time approach, and your
        code needs to request thousands of items, the sendBatchRequests() method is the one to use.

        >>> # Request info for 3 locations.
        >>> # Single-request URL is GET /Publish/v2/GetLocationData?LocationIdentifer=loc1
        >>> # Operation name is "LocationDataServiceRequest"
        >>> requests = [{'LocationIdentifier': 'Loc1'}, {'LocationIdentifier': 'Loc2'}, {'LocationIdentifier': 'Loc3'}]
        >>> responses = client.publish.send_batch_requests("LocationDataServiceRequest", requests)
        :param operation_name: The name of operation, from the AQTS Metadata page, to perform multiple times. NOT the route, but the operation name.
        :param requests: A list of individual request objects
        :param batch_size: Optional batch size (defaults to 100 requests per batch)
        :param verb: Optional HTTP verb of the operation (defaults to "GET")
        :return: A list of all the responses.
        """
        url = "/json/reply/{0}[]".format(operation_name)

        # Split the list into batches
        batched_requests = [requests[i:i + batch_size] for i in range(0, len(requests), batch_size)]

        # Get the response batches
        batched_responses = [self.post(url, json=batch, headers={'X-Http-Method-Override': verb}).json() for batch in batched_requests]

        # Return the flattened response list
        return [response for batch in batched_responses for response in batch]


class timeseries_client:
    """
    A client wrapper for AQUARIUS Time-Series REST API consumption.

    Each of the public REST APIs is exposed as a timeseries_session object:

    publish      => An authenticated session to the /AQUARIUS/Publish/v2 endpoint
    acquisition  => An authenticated session to the /AQUARIUS/Acquisition/v2 endpoint
    provisioning => An authenticated session to the /AQUARIUS/Provisioning/v1 endpoint

    >>> timeseries = timeseries_client('localhost', 'admin', 'admin')
    >>> timeseries.publish.get('/session').json()
    {u'Username': u'admin', u'Locale': u'en', u'Token': u'GWVheAEXYDkJrqKWxFA1vQ2', u'CanConfigureSystem': True, u'IpAddress': u'172.16.1.90'}

    Session resources will be automatically cleaned up if used in WITH statement.

    >>> with timeseries_client('localhost', 'admin', 'admin') as timeseries:
    ...   parameters = timeseries.publish.get('/GetParameterList').json()["Parameters"]
    ...   print "There are {0} parameters".format(len(parameters))
    ...
    >>> # The session will be disconnected now, even if an exception was thrown in the body of the WITH statement.
    """

    def __init__(self, hostname, username, password, verify=True):
        # Create the three endpoint sessions
        self.publish = TimeseriesSession(hostname, "/AQUARIUS/Publish/v2", verify=verify)
        self.acquisition = TimeseriesSession(hostname, "/AQUARIUS/Acquisition/v2", verify=verify)
        self.provisioning = TimeseriesSession(hostname, "/AQUARIUS/Provisioning/v1", verify=verify)

        # Authenticate once
        self.configure_proxy()
        self.connect(username, password)

        # Cache the server version
        versionSession = TimeseriesSession(hostname, "/AQUARIUS/apps/v1", verify=verify)
        self.serverVersion = versionSession.get('/version').json()["ApiVersion"]

    def __enter__(self):
        return self

    def __exit__(self, exception_type, exception_value, exception_traceback):
        self.disconnect()

    def process_exists(self, process_name):
        call = 'TASKLIST', '/FI', 'imagename eq %s' % process_name
        # use buildin check_output right away
        output = subprocess.check_output(call).decode()
        # check in last line for process name
        last_line = output.strip().split('\r\n')[-1]
        # because Fail message could be translated
        return last_line.lower().startswith(process_name.lower())

    def configure_proxy(self):
        if any(key in os.environ for key in ['PYTHON_DISABLE_FIDDLER', 'http_proxy', 'https_proxy']) or os.name != 'nt':
            return

        if self.process_exists('Fiddler.exe'):
            os.environ['http_proxy'] = os.environ['https_proxy'] = '127.0.0.1:8888'

    def connect(self, username, password):
        """
        Authenticates the session with AQUARIUS.

        All subsequent requests to any public endpoint will be authenticated using the stored session token.
        """
        token = self.publish.post('/session', json={'Username': username, 'EncryptedPassword': password}).text
        self.publish.set_session_token(token)
        self.acquisition.set_session_token(token)
        self.provisioning.set_session_token(token)

    def disconnect(self):
        """Destroys the authenticated session"""
        self.publish.delete('/session')

    def isVersionLessThan(self, sourceVersion, targetVersion):
        """
        Is the source version strictly less than the target version.

        :param sourceVersion: The source version, in dotted.string notation
        :param targetVersion: If None, the connected server version is used
        :return: True if the source version is strictly less than the target version
        """
        if targetVersion is None:
            targetVersion = self.serverVersion

        def createIntegerVector(versionText):

            if versionText == '0.0.0.0':
                # Force developer versions to be treated as latest-n-greatest
                versionText = '9999.99'

            versions = [int(i) for i in versionText.split('.')]

            if len(versions) > 0 and 14 <= versions[0] <= 99:
                # Adjust the leading component to match the 20xx.y release convention
                versions[0] += 2000

            return versions

        target = createIntegerVector(targetVersion)
        source = createIntegerVector(sourceVersion)

        for i in range(len(source)):
            if i >= len(target):
                return False

            if source[i] < target[i]:
                return True

            if source[i] > target[i]:
                return False

        return len(source) < len(target)

    def isServerVersionLessThan(self, targetVersion):
        return self.isVersionLessThan(self.serverVersion, targetVersion)

    def iso8601(self, datetime):
        """Formats the datetime object as an ISO8601 timestamp"""
        return pyrfc3339.generate(datetime, microseconds=True)

    def datetime(self, text):
        """Parses the ISO8601 timestamp to a standard python datetime object"""
        return pyrfc3339.parse(text)

    def coerceQueryTime(self, querytime):
        """Coerces the timevalue into a best possible query time format"""
        if isinstance(querytime, datetime):
            if querytime.tzinfo is None:
                # Format naive date times as a local time
                return querytime.strftime('%Y-%m-%d %H:%M:%S.%f')
            else:
                # Format unambiguous times as ISO8061
                return self.iso8601(querytime)

        # Otherwise return the value as-is and hope for the best
        return querytime

    def toJSV(self, item):
        """
        Converts non-scalar GET request parameters into JSV format.

        Lists must be comma separated
        Dictionaries must be in {Name1:Value1,Name2:Value2} format

        https://github.com/ServiceStack/ServiceStack.Text/wiki/JSV-Format
        :param item: item to be converted into JSV format
        :return: the JSV representation of the item
        """
        if isinstance(item, list):
            # Concatenate all values with commas
            return '[' + ','.join([str(self.toJSV(i)) for i in item]) + ']'
        if isinstance(item, dict):
            # Concatenate all the name/value pairs
            return '{' + ','.join([k+':'+str(self.toJSV(item[k])) for k in item.keys()]) + '}'

        return item

    def getTimeSeriesUniqueId(self, timeSeriesIdentifier):
        """Gets the unique ID of a time-series"""
        parts = timeSeriesIdentifier.split('@')

        if len(parts) < 2:
            return timeSeriesIdentifier

        location = parts[1]

        # Get the descriptions from the location
        try:
            descriptions = self.publish.get(
                '/GetTimeSeriesDescriptionList', params={'LocationIdentifier': location}).json()[
                "TimeSeriesDescriptions"]
        except requests.exceptions.HTTPError as e:
            raise LocationNotFoundException(location)

        matches = [d for d in descriptions if d['Identifier'] == timeSeriesIdentifier]

        if len(matches) != 1:
            raise TimeSeriesNotFoundException(timeSeriesIdentifier)

        return matches[0]['UniqueId']

    def getLocationIdentifier(self, timeSeriesOrRatingModelIdentifier):
        """Extracts the location identifier from a 'Parameter.Label@Location' time-series or rating model identifier"""
        parts = timeSeriesOrRatingModelIdentifier.split('@')

        if len(parts) < 2:
            raise LocationNotFoundException(timeSeriesOrRatingModelIdentifier)

        return parts[1]

    def getLocationData(self, identifier):
        """Gets the location data"""
        return self.publish.get(
            "/GetLocationData", params={'LocationIdentifier': identifier}).json()

    def getLocationUniqueId(self, locationIdentifier):
        """Gets the location unique ID for the location"""
        if re.match('^[0-9a-f]{32}$', locationIdentifier):
            # Return existing GUIDs as-is
            return locationIdentifier

        locationData = self.getLocationData(locationIdentifier)

        return locationData['UniqueId']

    def getRatings(self, locationIdentifier, queryFrom=None, queryTo=None, inputParameter=None, outputParameter=None):
        ratingModelDescriptions = self.publish.get(
            "/GetRatingModelDescriptionList",
            params={
                'LocationIdentifier': locationIdentifier,
                'QueryFrom': self.coerceQueryTime(queryFrom),
                'QueryTo': self.coerceQueryTime(queryTo),
                'InputParameter': inputParameter,
                'OutputParameter': outputParameter
            }).json()['RatingModelDescriptions']

        # TODO: Get the rating curves in effect during those times
        return ratingModelDescriptions

    def getRatingModelOutputValues(self, ratingModelIdentifier, inputValues, effectiveTime=None, applyShifts=None):
        return self.publish.get(
            "/GetRatingModelOutputValues",
            params={
                'RatingModelIdentifier': ratingModelIdentifier,
                'InputValues': inputValues,
                'EffectiveTime': self.coerceQueryTime(effectiveTime),
                'ApplyShifts': applyShifts
            }).json()['OutputValues']

    def getFieldVisits(self, locationIdentifier, queryFrom=None, queryTo=None, activityType=None):
        fieldVisitDescriptions = self.publish.get(
            "/GetFieldVisitDescriptionList",
            params={
                'LocationIdentifier': locationIdentifier,
                'QueryFrom': self.coerceQueryTime(queryFrom),
                'QueryTo': self.coerceQueryTime(queryTo),
                'ActivityType': activityType
            }).json()['FieldVisitDescriptions']

        # TODO: Fetch details for each visit
        return fieldVisitDescriptions

    def getTimeSeriesDescriptions(self, locationIdentifier=None, parameter=None, publish=None, computationIdentifier=None, computationPeriodIdentifier=None, extendedFilters=None):
        return self.publish.get(
            "/GetTimeSeriesDescriptionList",
            params={
                'LocationIdentifier': locationIdentifier,
                'Parameter': parameter,
                'Publish': publish,
                'ComputationIdentifier': computationIdentifier,
                'ComputationPeriodIdentifier': computationPeriodIdentifier,
                'ExtendedFilters': self.toJSV(extendedFilters)
            }).json()['TimeSeriesDescriptions']

    def getTimeSeriesData(self, timeSeriesIds, queryFrom=None, queryTo=None, outputUnitIds=None, includeGapMarkers=None):
        if isinstance(timeSeriesIds, list):
            timeSeriesIds = [self.getTimeSeriesUniqueId(ts) for ts in timeSeriesIds]
        else:
            timeSeriesIds = self.getTimeSeriesUniqueId(timeSeriesIds)

        return self.publish.get(
            "/GetTimeSeriesData",
            params={
                'TimeSeriesUniqueIds': self.toJSV(timeSeriesIds),
                'TimeSeriesOutputUnitIds': self.toJSV(outputUnitIds),
                'QueryFrom': self.coerceQueryTime(queryFrom),
                'QueryTo': self.coerceQueryTime(queryTo),
                'IncludeGapMarkers': includeGapMarkers
            }).json()

    def getTimeSeriesCorrectedData(self, timeSeriesIdentifier, queryFrom=None, queryTo=None, getParts=None, includeGapMarkers=None):
        return self.publish.get(
            "/GetTimeSeriesCorrectedData",
            params={
                'TimeSeriesUniqueId': self.getTimeSeriesUniqueId(timeSeriesIdentifier),
                'QueryFrom': self.coerceQueryTime(queryFrom),
                'QueryTo': self.coerceQueryTime(queryTo),
                'GetParts': getParts,
                'IncludeGapMarkers': includeGapMarkers
            }).json()

    def getReportList(self):
        """Gets all the generated reports on the system"""
        return self.publish.get("/GetReportList").json()['Reports']

    def deleteReport(self, reportUniqueId):
        """Deletes the generated report from the system"""
        self.acquisition.delete("/attachments/reports/" + reportUniqueId)

    def uploadExternalReport(self, locationUniqueId, pathToFile, title, deleteDuplicateReports=False):
        """Uploads a file as external report to the given location"""
        if deleteDuplicateReports:
            # Get the current reports
            for r in [r for r in self.getReportList() if r['Title'] == title and not r['IsTransient']]:
                self.deleteReport(r['ReportUniqueId'])

        return self.acquisition.post("/locations/" + locationUniqueId + "/attachments/reports",
                                     params={'Title': title},
                                     files={'file': open(pathToFile, 'rb')}).json()

    def uploadFieldVisit(self, locationUniqueId, pathToFile):
        """Uploads a file as a field visit to the given location"""
        return self.acquisition.post("/locations/" + locationUniqueId + "/visits/upload/plugins",
                                     files={'file': open(pathToFile, 'rb')}).json()
